### DISCUSSION TITLE
Practice Problem Solutions

### DISCUSSION THREAD

**Discussion by rparakhiya** [tark-ignite/codegenie-2023#3] [Show and tell]
Solved practice problems?

Upload it to your github repository and share the link with us here.

**Comment by Saurav-Sutaria** [comment]
https://github.com/Saurav-Sutaria/codegenie-2023

Here's my solutions of the practise problems.

**Reply by rparakhiya** [reply]
Pretty quick! Nice! 👏

**Reply by Saurav-Sutaria** [reply]
Thank you Sir 😊

**Comment by vimal-parakhiya** [comment]
Saurav, could you please also help by creating skeleton code in CPP
(similar to Java skeleton) for benefit of other students? You may submit a
PR with skeleton code.

**Reply by Saurav-Sutaria** [reply]
Sure

**Reply by Saurav-Sutaria** [reply]
Sir, I have added the skeleton and created a PR #5 ,
Please verify it and do let me know if it requires any modifications.

**Reply by vimal-parakhiya** [reply]
Thanks, this is approved and merged.

**Comment by uttamsutariya** [comment]
Here is solution of practice problems in java. 

[https://github.com/uttamsutariya/codegenie-practice-problems](https://github.com/uttamsutariya/codegenie-practice-problems)

**Reply by vimal-parakhiya** [reply]
Nice! Good Work! I liked the usage of Comparator along with Student class.

**Comment by alifiyakapasi07** [comment]
Here's my solutions of the practice problems in python.

https://github.com/alifiyakapasi07/CodeGenie_2023

**Reply by vimal-parakhiya** [reply]
Good attempt Alifiya!
I believe, you may want to revisit the solutions of StudentScreener and GuessTheNumber.

StudentScreener: For the same example (given in problem statement), if I just change the eligibility_percentile to 80. Ideally, only one student (Kartik) should be qualified for the scholarship. I believe, your solution returns multiple students in this case. However, I liked the usage of zip function to create tuples (correlated data set).

GuessTheNumber: If final number is 2500, the expected output should be 5 (as explained in the problem statement). I believe your solution returns 111 in this case.

**Reply by alifiyakapasi07** [reply]
Okay Sir , I will check this

**Reply by alifiyakapasi07** [reply]
I have updated my code , Kindly review

**Reply by vimal-parakhiya** [reply]
`GuessTheNumber`: Looks good 👍 

`StudentScreener`: Seems to be generating incorrect output for below scenario. You may want to check that.

```
Students: ["Kartik", "Devang",   "Pari",  "Ketan", "Sheetal", "Darshana",  "Mohan"]
Score:      [     760,      300,      750,      800,       680,        790,      640]
Result:     ["Passed", "Failed", "Passed", "Passed",  "Passed",   "Passed", "Passed"]
Required Percentile: 50 

Actual Output: Kartik,Ketan,Darshana
Expected Output: Ketan,Darshana,Kartik

```

```
Students: ["Kartik", "Devang",   "Pari",  "Ketan", "Sheetal", "Darshana",  "Mohan"]
Score:      [     760,      300,      800,      750,       680,        790,      640]
Result:     ["Passed", "Passed", "Passed", "Passed",  "Passed",   "Passed", "Passed"]
Required Percentile: 50 

Actual Output: Kartik,Pari,Darshana
Expected Output: Pari,Darshana,Kartik
```

**Thoughts:**  It might be helpful to validate the solution, using multiple test data, because quite often, a problem description contains only representative examples/scenarios (to clarify the problem statement).

**Comment by sachin027** [comment]
Here's the solution of all practice problems solution in java .
https://github.com/sachin027/CodeGenie-2023

**Reply by vimal-parakhiya** [reply]
Nice solutions Sachin!

It seems that the `StudentScreener` solution is coded for specific example only. Please note that, during actual programming round, there will be multiple test-cases and your solution should work for all of them. So, you may want to review and change it accordingly.

**Reply by sachin027** [reply]
yes sir , thanks for bringing it to my attention  . 

i have updated my code using HashMap . kindly review and let me know if still have any mistake .

**Reply by vimal-parakhiya** [reply]
Looks good and while using hashmap + Arrays.sort makes it concise, your earlier solution was also correct, it was just needed to converted into generic solution. Just for your information.

**Comment by aminlodhiya** [comment]
Here's My Java Solution of all the Problems. 
https://github.com/Mahmadamin08/CodeGenie_2023

Thank you,

**Reply by vimal-parakhiya** [reply]
Good Solutions!

`GuessTheNumber`: I believe, it does not return expected output for below inputs,

Final Number: 100
Operations: ["X ^ 2"]
Expected Output: 10

**Hint:** Watch out for floating point operations, where applicable.

**Reply by aminlodhiya** [reply]
Ok! Got it.
i Updated my code. 
please verify and let me know if still have any Mistake..

Thank you,

**Reply by vimal-parakhiya** [reply]
:thumbsup: Looks good!

**Comment by togadiyamit** [comment]
here's my problems solution using java
https://github.com/togadiyamit/Practice-problems

**Reply by vimal-parakhiya** [reply]
Good solutions! 👍

**Comment by khunt-miral** [comment]
My solution repo

https://github.com/khunt-miral/codegenie-2023/tree/practice-problems/practice-problems/skeletons

**Reply by vimal-parakhiya** [reply]
👍 Nicely used StudentClass and Comparator. Having a separate method to calculate `percentile` also improves readability!

**Comment by JayDabhi240** [comment]
https://github.com/Dabhi-Jay/TarkTechnologies_Practise_Problems

My solutions for given three problems, although it is partial for problem 3(Guess the number)

**Reply by vimal-parakhiya** [reply]
Looks good!

Few suggestions,
1. During actual coding round (tomorrow), you do not have to read data using cin/stdin. All inputs will be given as parameters to the function and you will have to just implement that function and return correct answer. You may also want to refer the video shared in yesterday's mail for reference.

2. Converting string to int - you may want to use string class and substr function to pick the sub-string representing the number and then you can use atoi function to convert that string (of numbers) to int. This will save time as you do not have to iterate over each character and then converting into a number.

**Reply by JayDabhi240** [reply]
Got it! 
Thank you.

**Comment by brindadavda** [comment]
https://github.com/brindadavda/codegenie-practice-problems

My solution repo in java of all the given problem.

**Comment by Yash-Daxini** [comment]
Final Round problem Till Milestone - 1 

Repository Link :- [Train Reservation System](https://github.com/Yash-Daxini/tark-codegenie-2023-final)

**Reply by vimal-parakhiya** [reply]
@Yash-Daxini while our team reviews the code, you may want to validate the solution against full set of test cases, as mentioned [here](https://bit.ly/codegenie-2023-all-tests). Please keep us posted once you validate it by executing all test cases (Level 4 and Level 5)

**Reply by Yash-Daxini** [reply]
Ok sir

**Reply by Yash-Daxini** [reply]
Sir , I validated level 4 and level 5 and it matches with output.

**Reply by vimal-parakhiya** [reply]
@Yash-Daxini  code is readable and looks clean. However, could you validate the solution for below inputs?

```
1
17098 Rajkot-0 Mumbai-750
17098 S1-5 S2-10
Rajkot Mumbai 2022-03-25 SL 15 (Should be successful)
Rajkot Mumbai 2022-03-30 SL 15 (Should be successful)
Rajkot Mumbai 2022-03-25 SL 2 (Should display - No Seats Available)
```

**Reply by Yash-Daxini** [reply]
Solution didn't work for the above test case . I'll fix it .

**Reply by Yash-Daxini** [reply]
Now, solution satisfies all the given testcases.

Please, review it.

Link :- [Trains Reservation System](https://github.com/Yash-Daxini/tark-codegenie-2023-final)

**Reply by vimal-parakhiya** [reply]
Sure @Yash-Daxini we will check this and keep you posted. While we review this, you may want to try with next milestone.

**Reply by Yash-Daxini** [reply]
ok sir .
